File: a3.jl
trees = open("a3.txt") do file
  hcat([map((t) -> t == '#' ? 1 : 0, collect(line)) for line in eachline(file)]...)
end
width, height = size(trees)

function slope(h, v)
  Iterators.zip(
    Iterators.accumulate((x, i) -> mod1(x+i, width), Iterators.repeated(h); init=1-h),
    1:v:height
  )
end

function ride(h,v)
  sum(trees[CartesianIndex.(slope(h, v))])
end

#part1
println("$(ride(3,1))")

part2 = ride(1, 1) * ride(3, 1) * ride(5, 1) * ride(7, 1) * ride(1, 2)
println("$part2")

File: a3.dart
import 'dart:io';

class Ride {
  int horizontal;
  int vertical;
  Ride(this.horizontal, this.vertical);

  int go(List<List<int>> trees) {
    int width = trees[0].length;
    int hit = 0;
    int h = 0;
    int v = 0;
    do {
      hit += trees[v][h];
      h = (h + horizontal) % width;
      v += vertical;
    } while (v < trees.length);
    return hit;
  }
}

class A3 {
  List<List<int>> trees;
  A3(this.trees);

  int part1() {
    return new Ride(3, 1).go(trees);
  }

  int part2() {
    return new Ride(1, 1).go(trees) *
        new Ride(3, 1).go(trees) *
        new Ride(5, 1).go(trees) *
        new Ride(7, 1).go(trees) *
        new Ride(1, 2).go(trees);
  }
}

void main(List<String> args) {
  var input = new File('a3.txt');

  var trees = input
      .readAsLinesSync()
      .map((l) => l.split('').map((t) => t == '#' ? 1 : 0).toList())
      .toList();
  var solver = new A3(trees);
  stdout.writeln(solver.part1());
  stdout.writeln(solver.part2());
}
